### DISCUSSION TITLE
query_datasource_plus_xyz runs forever sometimes

### DISCUSSION THREAD

**Discussion by R2bEEaton** [Budibase/budibase#11291] [Help] [answered]
Hello! I'm having this weird issue that only seems to be affecting this page of my Budibase site on my Raspberry Pi devices, tested on Chrome and Firefox, on multiple Raspberry Pis.

![image](https://github.com/Budibase/budibase/assets/34921506/9c0ec644-2de1-4767-98e1-e469c37b91e6)

Sometimes the page loads fine in under a second. But most often, the query_datasource_plus_xyz request will keep getting run over and over. It always retrieves the response data successfully, so it's not like it's failing. The solution is to keep refreshing the entire page until it decides to load instantly. If it doesn't load quickly, it seems to be a lost cause already.

I wish I had more information to give, but I have no idea what could be causing this. The data sources on the top of the page always load instantly, and there are no nested data providers in the area that has the forever-spinning wheel GIF. Just a simple DataSource + Table combo. This page loads instantly on Chrome on Windows, on my phone, and everywhere else I've tried.

**Comment by ConorWebb96** [comment]
Hey @R2bEEaton,

- Are you self-host or cloud based?
- If you are self-host are you running this off the Raspberry Pi?
- What version of Budibase are you working with?
- Would we be able to get an app export? (Screenshots of your structure, e.g. the queries you have made and your component tree would be useful.)

**Reply by R2bEEaton** [reply]
Good morning, thanks for getting back!

* Self-hosted
* No, it is running elsewhere
* v2.7.14, scared to update #11052
* Sure, though I think this is the only relevant screenshot

![image](https://github.com/Budibase/budibase/assets/34921506/6239a57c-e2d0-48cf-a793-67de25e1abd5)
All the other data providers work just fine, and all the elements inside this one are just for display purposes. It's not the Interval element plugin, I've tried removing it.

Actually, now that I notice it, it looks like my query was expecting a date in MM/DD/YYYY format and I was giving it `{{ now }}`. Let me try giving it `{{ date now "M/D/YYYY" }}` specifically.

???? It works now on the Raspberry Pi. It doesn't do the request over and over again. Same data as it was showing on Windows before. I don't see why this could be the problem, because it was working fine when I visited the site on a Windows machine. The request response is identical, it's showing all the same data.

If I change it back to `{{ now }}` it breaks again, repeating the request every few seconds and never displaying the table.

**Reply by ConorWebb96** [reply]
Hey @R2bEEaton,

This is very strange, I have had a look a this and I wasn't able to replicate this. (One of our engineers also had a look and they could replicate it either.) Could we possibly get an app export so we can have a more in-depth look with your exact settings. (If you aren't comfortable sharing the current app shown in the screenshots, would it be possible to create a dummy app were this is occurring that you'd be fine with sharing.)

Thanks ahead of time!

**Reply by R2bEEaton** [reply]
Sure, sorry for the delay. I must admit, I had some trouble reproducing it and was unable to learn anything about why it happens. But it happens on this export: [githubexportimport-export-1690281613361.tar.gz](https://github.com/Budibase/budibase/files/12160125/githubexportimport-export-1690281613361.tar.gz)

It is referencing a custom query on a SQL server without any used bindings, which returns only 73 rows with 5 columns (not a lot of data). For your testing, I'm guessing you could replace it with your own SQL server table of similar size. On the Pi's, the query_datasource_plus_xyz request runs around 10 times (spanning 10 seconds) before displaying something on the screen. On a Windows machine, running MS Edge, it displays much quicker and with only one request.

Here's an export of the network tab on Chromium from the Raspeberry Pi, which can then be imported and viewed if you want:
[budi.har.zip](https://github.com/Budibase/budibase/files/12160161/budi.weberdisplay-pkg.com.zip)
![image](https://github.com/Budibase/budibase/assets/34921506/b1da78b6-dcc3-4c5b-b227-76bbf49168fa)

Let me know if you need any more information, I'm happy to provide it.

**Reply by ConorWebb96** [reply, marked as answer]
Hey @R2bEEaton,

Firstly, thank you very much for your exports. They were very helpful!

I got one of the engineers to have a look at this. The reason it was constantly loading/refreshing was due to the `{{ now }}` binding (Its not a static date, thus constantly changing/causes the queries to constantly reload). It caused the datasources to become constantly invalidated and thus kept triggering the refreshes to try and fix it. The fix here is to input a static date, you could use `{{ now }}` to output the current date/time to state then pass that in as the bindable value. 

I hope this explanation helps and makes sense.

**Reply by R2bEEaton** [reply]
Hey @ConorWebb96

Ohhhhhhh! This explanation makes total sense and explains other related problems I've been having. Thanks so much for this! Just out of curiosity though, was the engineer able to determine why this was happening for the Raspberry Pis (and only 10 times) and never for the Windows clients? I never had a problem like this on a Windows machine. At this point it's just curiosity. Problem solved!

Please thank the engineer for me!
